### PR TITLE
Validate genomic location input

### PR DESCRIPTION
**Pull request by leexgh**
This checks whether the genomic location input of the request doesn't contain any unexpected characters.

- [x] logic implemented
- [x] tests
- [ ] return 4x error on incorrect input instead of 5x error

### CODE REVIEW

File: web/src/main/java/org/cbioportal/genome_nexus/web/validation/GenomicLocationValidator.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package org.cbioportal.genome_nexus.web.validation;
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+// import org.cbioportal.genome_nexus.model.GenomicLocation;
```
Comment (onursumer) [inline web/src/main/java/org/cbioportal/genome_nexus/web/validation/GenomicLocationValidator.java]: we can remove this line

File: web/src/main/java/org/cbioportal/genome_nexus/web/validation/IsValidGenomicLocation.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target({ ElementType.TYPE, ElementType.METHOD, ElementType.FIELD, ElementType.PARAMETER })
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Constraint(validatedBy = { GenomicLocationValidator.class })
+public @interface IsValidGenomicLocation {
```
Comment (onursumer) [inline web/src/main/java/org/cbioportal/genome_nexus/web/validation/IsValidGenomicLocation.java]: Let's just name this `ValidGenomicLocation` instead

File: web/src/main/java/org/cbioportal/genome_nexus/web/EnsemblController.java [hunk 3]
```diff
@@ -3,6 +3,7 @@
 import io.swagger.annotations.Api;
 import io.swagger.annotations.ApiOperation;
 import io.swagger.annotations.ApiParam;
+import io.swagger.annotations.ExampleProperty;
```
Comment (onursumer) [inline web/src/main/java/org/cbioportal/genome_nexus/web/EnsemblController.java]: Please revert all changes made on this file, since it is not related to this PR

File: web/src/main/java/org/cbioportal/genome_nexus/web/AnnotationController.java [hunk 4]
```diff
@@ -33,32 +33,39 @@
...
 @PublicApi
 @RestController // shorthand for @Controller, @ResponseBody
 @CrossOrigin(origins="*") // allow all cross-domain requests
 @RequestMapping(value= "/")
+@Validated
 public class AnnotationController
 {
     private final VariantAnnotationService variantAnnotationService;
 
+
```
Comment (onursumer) [inline web/src/main/java/org/cbioportal/genome_nexus/web/AnnotationController.java]: remove the whitespace here

File: web/src/main/java/org/cbioportal/genome_nexus/web/AnnotationController.java [hunk 5]
```diff
@@ -33,32 +33,39 @@
...
     @Autowired
     public AnnotationController(VariantAnnotationService variantAnnotationService)
     {
         this.variantAnnotationService = variantAnnotationService;
+
```
Comment (onursumer) [inline web/src/main/java/org/cbioportal/genome_nexus/web/AnnotationController.java]: remove the whitespace here

File: web/pom.xml [hunk 6]
```diff
@@ -58,13 +58,13 @@
     <dependency>
       <groupId>io.springfox</groupId>
       <artifactId>springfox-swagger2</artifactId>
-      <version>2.7.0</version>
+      <version>2.9.2</version>
```
Comment (onursumer) [inline web/pom.xml]: this is nice to have, but maybe do this in another PR?

File: web/pom.xml [hunk 7]
```diff
@@ -58,13 +58,13 @@
...
       <scope>compile</scope>
     </dependency>
     <dependency>
       <groupId>io.springfox</groupId>
       <artifactId>springfox-swagger-ui</artifactId>
-      <version>2.7.0</version>
+      <version>2.9.2</version>
```
Comment (onursumer) [inline web/pom.xml]: this is nice to have, but maybe do this in another PR?

**Comment by leexgh** [comment]
@onursumer I have updated, thanks for the review!

File: web/src/test/java/org/cbioportal/genome_nexus/web/validation/GenomicLocationValidatorTest.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    public void testMyVariantInfoAnnotation()
+    {
+        String[] genomicLocation = {
+            "7,140453136,140453136,A,ATGC-T", // true
+            "A,37880220,37880220,T,C", // false
+            "7,-1,37879794,G,C", // false
+            "12,25398284,-2,CC,GA", // false
+            "12,25380271,25380271,12,T", // false
+            "10,89692940,89692940,C,Q", // false
+            "X,89692940,89692940,C," // false
```
Comment (onursumer) [inline web/src/test/java/org/cbioportal/genome_nexus/web/validation/GenomicLocationValidatorTest.java]: Let's add one more test for malformed input, something like `X_89692940_89692940_C_T`

File: web/src/main/java/org/cbioportal/genome_nexus/web/validation/GenomicLocationValidator.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+public class GenomicLocationValidator implements ConstraintValidator<ValidGenomicLocation, String> {
+    
+    public void initialize(ValidGenomicLocation constraintAnnotation) {
+    }
+
+    public boolean isValid(String genomicLocation, ConstraintValidatorContext context) 
+    {
+        String[] genomicLocationSplit = genomicLocation.split(",");
+        if (genomicLocationSplit.length <5){
```
Comment (onursumer) [inline web/src/main/java/org/cbioportal/genome_nexus/web/validation/GenomicLocationValidator.java]: always leave a whitespace after `<`, and also between `)` and `{`

File: web/src/main/java/org/cbioportal/genome_nexus/web/validation/GenomicLocationValidator.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+            return false;
+        }
+        else{
```
Comment (onursumer) [inline web/src/main/java/org/cbioportal/genome_nexus/web/validation/GenomicLocationValidator.java]: always leave a whitespace between `else` and `{`

File: web/src/main/java/org/cbioportal/genome_nexus/web/validation/GenomicLocationValidator.java [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+            boolean result = true;
+            // chr [1,24] or X or Y
+            result &= genomicLocationSplit[0].matches("X") || genomicLocationSplit[0].matches("Y") || genomicLocationSplit[0].matches("[0-9]+") && Integer.valueOf(genomicLocationSplit[0]) >= 1 && Integer.valueOf(genomicLocationSplit[0]) <= 24;
```
Comment (onursumer) [inline web/src/main/java/org/cbioportal/genome_nexus/web/validation/GenomicLocationValidator.java]: you should break long lines of conditions into multiple lines for better readability, for example something like this:

```suggestion
            result &= genomicLocationSplit[0].matches("X") || genomicLocationSplit[0].matches("Y") ||
             (genomicLocationSplit[0].matches("[0-9]+") && Integer.valueOf(genomicLocationSplit[0]) >= 1 && Integer.valueOf(genomicLocationSplit[0]) <= 24);
```

File: web/src/main/java/org/cbioportal/genome_nexus/web/AnnotationController.java [hunk 12]
```diff
@@ -33,32 +33,37 @@
...
+@Validated
 public class AnnotationController
 {
     private final VariantAnnotationService variantAnnotationService;
 
     @Autowired
     public AnnotationController(VariantAnnotationService variantAnnotationService)
     {
-        this.variantAnnotationService = variantAnnotationService;
+        this.variantAnnotationService = variantAnnotationService;
```
Comment (onursumer) [inline web/src/main/java/org/cbioportal/genome_nexus/web/AnnotationController.java]: can you revert this line, there is unnecessary whitespace at the end?

**Comment by leexgh** [comment]
@onursumer Updated, thanks for the review!
